WordPress/gutenberg
# Migrate post editor preferred style variations to preferences store

## What?
Addresses some of https://github.com/WordPress/gutenberg/issues/31965

Migrates the post editor's preferred style variations data to the preferences store.

## Why?
The roadmap is detailed in #31965.

The goal is to use the preferences package for storing any persisted preference state, removing persistence from all other packages.

This will:
- Allow for cross editor settings
- Make it easier to implement persisting preferences to the database in the future
- Eventually allow us to deprecate the 'persistence' store plugin
- Allow us to stop using `registerStore` in favor of `register`.

## How?
`preferredStyleVariations` are a little tricky. The way this system works now isn't ideal because:
- it only works in the post editor
- it uses editor settings [as a mechanism for dependency injection](https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-post/src/editor.js#L117-L120), which now seems like an anti-pattern, but I think was done because there was no other solution at the time. 
- the feature is currently half disabled pending better integration with global styles (since #34522)

I could've fixed the first two things by giving it a revamp, but given the last point I've decided to leave everything working in the same way. The discussion in #34522 seems to point to this working very differently in the near future.

## Testing Instructions
To test, you'll need to temporarily re-enable the feature by commenting out these three lines:
https://github.com/WordPress/gutenberg/blob/680d1aa4adae9b8d2fa895a30b3dfd48f40837d9/packages/block-editor/src/components/default-style-picker/index.js#L54-L56

Then:
1. Delete your WP localstorage data
2. Using `trunk` open the post editor and dismiss the welcome guide
3. Add a quote and a table block and use the 'Default Style' dropdown to select a non-default style
4. Checkout this branch and rebuild
5. Reload the post editor
6. The welcome guide should not show
7. Add a quote and a table again, the selected default style should still be applied (the localstorage migration worked)
9. Change quote back to 'Not set'
10. Reload
11. Add a quote and a table again, the quote should be back to using its default (changes made using the preference store are persisted correctly)
